pkg/authentication/scram/scram.go

```diff
@@ -47,9 +78,250 @@ func EnsureAgentSecret(getUpdateCreator secret.GetUpdateCreator, secretNsName ty
 	}
 
+	if err := secret.CreateOrUpdate(secretGetUpdateCreateDeleter, agentSecret); err != nil {
+		return automationconfig.NOOP(), err
+	}
+
 	return automationConfigModification(
 		string(agentSecret.Data[AgentPasswordKey]),
-		string(agentSecret.Data[AgentKeyfileKey]),
-		[]automationconfig.MongoDBUser{},
-	), getUpdateCreator.UpdateSecret(agentSecret)
+		string(agentSecret.Data[AgentKeyfileKey]), desiredUsers,
+	), nil
+}
+
+// User defines the required methods for a MongoDB user to provide
+// in order to generate ScramCredentials
+type User interface {
+	GetPasswordSecretKey() string
+	GetPasswordSecretName() string
+	GetUserName() string
+}
+
+// ensureScramCredentials will ensure that the ScramSha1 & ScramSha256 credentials exist and are stored in the credentials
+// secret corresponding to user of the given MongoDB deployment.
+func ensureScramCredentials(getUpdateCreator secret.GetUpdateCreator, user User, mdbNamespacedName types.NamespacedName) (scramcredentials.ScramCreds, scramcredentials.ScramCreds, error) {
+
+	password, err := secret.ReadKey(getUpdateCreator, user.GetPasswordSecretKey(), types.NamespacedName{Name: user.GetPasswordSecretName(), Namespace: mdbNamespacedName.Namespace})
+	if err != nil {
+		// if the password is deleted, that's fine we can read from the stored credentials that were previously generated
+		if errors.IsNotFound(err) {
+			zap.S().Debugf("password secret was not found, reading from credentials from secret/%s", scramCredentialsSecretName(mdbNamespacedName.Name, user.GetUserName()))
+			return readExistingCredentials(getUpdateCreator, mdbNamespacedName, user.GetUserName())
+		}
+		return scramcredentials.ScramCreds{}, scramcredentials.ScramCreds{}, err
+	}
+
+	// we should only need to generate new credentials in two situations.
+	// 1. We are creating the credentials for the first time
+	// 2. We are changing the password
+	needToGenerateNewCredentials, err := needToGenerateNewCredentials(getUpdateCreator, user.GetUserName(), mdbNamespacedName, password)
+	if err != nil {
+		return scramcredentials.ScramCreds{}, scramcredentials.ScramCreds{}, err
+	}
+
+	// there are no changes required, we can re-use the same credentials.
+	if !needToGenerateNewCredentials {
+		zap.S().Debugf("Credentials have not changed, using credentials stored in: secret/%s", scramCredentialsSecretName(mdbNamespacedName.Name, user.GetUserName()))
+		return readExistingCredentials(getUpdateCreator, mdbNamespacedName, user.GetUserName())
+	}
+
+	// the password has changed, or we are generating it for the first time
+	sha1Creds, sha256Creds, err := generateScramShaCredentials(user.GetUserName(), password)
+	if err != nil {
+		return scramcredentials.ScramCreds{}, scramcredentials.ScramCreds{}, err
+	}
+
+	// create or update our credentials secret for this user
+	zap.S().Debugf("Generating new credentials and storing in secret/%s", scramCredentialsSecretName(mdbNamespacedName.Name, user.GetUserName()))
```

Comment:
the log here does not seems to be in-line with what's going on, the credentials have been generated already